### PR TITLE
CI Profiling precursor changes

### PR DESCRIPTION
**Pull request by tomponline**
This PR:

1. Moves the building of the LXD binaries for the system tests into the code test stage and uses GH artifacts system in order to avoid building the LXD binaries multiple times for each concurrent runner.
2. Adds functionality to build LXD binaries with code coverage support and uploads them to the GH artifact system for collection later (this will come in a future PR).
3. Various Makefile improvements and cleanups.

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -242,20 +289,10 @@ jobs:
...
-          go mod tidy -go="${GOMIN}"
-
-          DOC_GOMIN="$(sed -n 's/^LXD requires Go \([0-9.]\+\) .*/\1/p' doc/requirements.md)"
-          [ "${GOMIN}" = "${DOC_GOMIN}" ]
-
       - name: Install dependencies
         run: |
           set -eux
           sudo add-apt-repository ppa:ubuntu-lxc/daily -y --no-update
-          sudo add-apt-repository ppa:dqlite/dev -y --no-update
```
Comment (simondeziel) [inline .github/workflows/tests.yml:258]: If we are not compiling LXD here, do we need the `ubuntu-lxc/daily` PPA and the `liblxc-dev` and other `-dev` packages?

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -391,8 +421,15 @@ jobs:
           set -eux
           chmod +x ~
           echo "root:1000000:1000000000" | sudo tee /etc/subuid /etc/subgid
+          mkdir -p "${GOCOVERDIR}"
           cd test
-          sudo --preserve-env=PATH,GOPATH,GITHUB_ACTIONS,LXD_VERBOSE,LXD_BACKEND,LXD_CEPH_CLUSTER,LXD_CEPH_CEPHFS,LXD_CEPH_CEPHOBJECT_RADOSGW,LXD_OFFLINE,LXD_SKIP_TESTS,LXD_REQUIRED_TESTS, LXD_BACKEND=${{ matrix.backend }} ./main.sh ${{ matrix.suite }}
+          sudo --preserve-env=PATH,GOPATH,GOCOVERDIR,GITHUB_ACTIONS,LXD_VERBOSE,LXD_BACKEND,LXD_CEPH_CLUSTER,LXD_CEPH_CEPHFS,LXD_CEPH_CEPHOBJECT_RADOSGW,LXD_OFFLINE,LXD_SKIP_TESTS,LXD_REQUIRED_TESTS, LXD_BACKEND=${{ matrix.backend }} ./main.sh ${{ matrix.suite }}
```
Comment (simondeziel) [inline .github/workflows/tests.yml:426]: In the `System` tests, do we still need Go tooling? Maybe to assist in collecting the coverage info?

**Review by simondeziel — APPROVED** [review]
LGTM thanks! There might be some things to further remove but those can wait till later.

One thing we lose with that is we are now only testing/building on Go 1.22. That said, maybe you wanted to have that done first and then we add another step to the `Code` job later on.

File: .github/workflows/tests.yml (hunk 1 above)
Comment (tomponline) [reply .github/workflows/tests.yml:258]: for now we do as LXD links against it dynamically

File: .github/workflows/tests.yml (hunk 2 above)
Comment (tomponline) [reply .github/workflows/tests.yml:426]: yeah we do for now at least because there are several go programs built during the tests to help run them

**Comment by tomponline** [comment]
> One thing we lose with that is we are now only testing/building on Go 1.22. That said, maybe you wanted to have that done first and then we add another step to the `Code` job later on.

Note, thanks for highlighting.

**Pull request merged**